namenode/main.go

```diff
@@ -250,11 +251,101 @@ func (server *NameNode) ReportBlock(blockReport common.BlockReport, success *boo
+func (server *NameNode) replicationMonitor() {
+	slog.Info("Replication monitor setup")
+
+	for {
+		time.Sleep(common.REPLICATION_MONITOR_INTERVAL) // Check periodically
+		slog.Info("Replication monitor running")
+
+		for fileName, inode := range server.inodes {
+			inode.rwlock.RLock()
+			for blockIndex, storageInfo := range inode.storageInfo {
+				if len(storageInfo.DataNodes) < common.BLOCK_REPLICATION && storageInfo.LatestVersion > 0 {
+					// Under-replicated block
+					slog.Info("Under-replicated block", "fileName", inode.fileAttributes, "blockIndex", blockIndex)
+
+					// Pick an alive datanode holding the block to replicate
+					var aliveStorageNodes []string
+					for _, storageNode := range storageInfo.DataNodes {
+						if server.datanodes[storageNode].isAlive {
+							aliveStorageNodes = append(aliveStorageNodes, storageNode)
+						}
+					}
+					if len(aliveStorageNodes) == 0 {
+						slog.Error("No alive storage nodes available for replication", "fileName", fileName, "blockIndex", blockIndex)
+						continue
+					}
+					dataNodeEndpoint := aliveStorageNodes[rand.Intn(len(aliveStorageNodes))]
+
+					// Pick datanodes that don't hold the block to replicate
+					replicaEndpoints := server.pickDatanodes(
+						uint(common.BLOCK_REPLICATION-len(storageInfo.DataNodes)),
+						storageInfo.DataNodes,
+					)
+
+					// Schedule replication: send a replication request to the datanode
+					go func(fileName string, blockIndex uint, replicaEndpoints []string, version uint, dataNodeEndpoint string) {
+						server.datanodeRWLock.RLock()
+						defer server.datanodeRWLock.RUnlock()
+
+						createReplicationRequest := common.CreateReplicationRequest{
+							FileName:         fileName,
+							BlockIndex:       blockIndex,
+							Version:          version,
+							ReplicaEndpoints: replicaEndpoints,
+						}
+
+						dataNodeClient, err := rpc.DialHTTP("tcp", dataNodeEndpoint)
+						if err != nil {
+							slog.Error("dialing datanode failed", "datanode endpoint", dataNodeEndpoint, "error", err)
+							return
+						}
+
+						replicateBlockTimeout :=
+							time.Millisecond*(common.BLOCK_SIZE*1024/common.NETWORK_BANDWIDTH)*3 +
+								common.RPC_TIMEOUT
+
+						slog.Info("Send replication request", "fileName", fileName, "blockIndex", blockIndex, "replicaEndpoints", replicaEndpoints, "dataNodeEndpoint", dataNodeEndpoint)
+						slog.Debug("create", slog.Any("createReplicationRequest", createReplicationRequest))
+
+						var unused bool
+
+						asyncRpcCall := dataNodeClient.Go("DataNode.CreateReplication", createReplicationRequest, &unused, nil)
+
+						select {
+						case <-asyncRpcCall.Done:
+							if asyncRpcCall.Error != nil {
+								slog.Error("Error during Create Replication request", "error", asyncRpcCall.Error)
+							} else {
+								// Update the storage info
+								inode.rwlock.Lock()
+								storageInfo.DataNodes = append(storageInfo.DataNodes, replicaEndpoints...)
```

Comment:
Let me mark this and consider it later. I got a warning "loop variable storageInfo captured by func literal" for `storageInfo`

Reply:
Thanks! I'm not sure what should be the best practice here

Reply:
Lambda functions capture variables by reference, but what we want is capture by value. Thus, we have to pass this as a parameter of the lambda function.